File: 0046-Permutations/0046-Permutations.go
func permutearr(list *[][]int, temp []int, nums []int) {
	if len(temp) == len(nums) {
		newTemp := make([]int, len(temp))
		copy(newTemp, temp)
		*list = append(*list, newTemp)
		return
	}
	for i := 0; i < len(nums); i++ {
		if contains(temp, nums[i]) {
			continue
		}
		temp = append(temp, nums[i])
		permutearr(list, temp, nums)
		temp = temp[:len(temp)-1]
	}
}

func contains(arr []int, target int) bool {
	for _, val := range arr {
		if val == target {
			return true
		}
	}
	return false
}

func permute(nums []int) [][]int {
	var result [][]int
	permutearr(&result, []int{}, nums)
	return result
}

File: 0046-Permutations/0046-Permutations.java
class Solution {
    void per(List<List<Integer>> list,List<Integer> temp,int[] nums){
        if(temp.size() == nums.length) {
            list.add(new ArrayList<>(temp));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (temp.contains(nums[i])) continue;
            temp.add(nums[i]);
            per(list, temp, nums);
            temp.remove(temp.size() - 1);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        per(list, new ArrayList<Integer>(), nums);
        return list;
        
    }
}
